Fix using the built-in DB with OpenShift

megian opened the pull request:
Security context parameters have changed in the helm chart.

Adding an additional test, because this test case is not covered.

## Checklist

- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

megian commented:
Looks like the shmVolume and volume permissions are set correctly by OpenShift:
```
$ ls -l /dev/shm
total 16
-rw-------. 1 1000980000 root 14992 Oct 26 10:21 PostgreSQL.901364491
$ mount | grep shm
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,context="system_u:object_r:container_file_t:s0:c25,c31",size=65536k)
```

simu commented:
Might make sense to figure out if it's possible to configure sensible `securityPolicy` which doesn't cause PSA warnings, cf. https://kb.vshn.ch/oc4/explanations/pod_security.html

ccremer commented on docs/modules/ROOT/pages/how-tos/upgrade-9.x-to-10.x.adoc:
```diff
@@ -7,6 +7,10 @@ This guide describes the steps to perform an upgrade of the component from versi
 * `postgresql_helm_values` authentication parameters have been moved to `postgresql_helm_values.auth` reflecting the https://docs.bitnami.com/kubernetes/infrastructure/postgresql/administration/upgrade/#to-1100[Helm chart upgrade] to 11.
+* `postgresql_helm_values.securityContext.enabled` changed to `postgresql_helm_values.primary.securityContext.enabled`.
+* `postgresql_helm_values.containerSecurityContext.enabled` changed to `postgresql_helm_values.primary.containerSecurityContext.enabled`.
+* `postgresql_helm_values.volumePermissions.securityContext.runAsUser` has no direct equivalent and therefore the volume permissions setup has to be disabled entirly `postgresql_helm_values.volumePermissions.enabled`.
```
What does `disabled` mean? setting to `false` or removing from catalog?
```suggestion
* `postgresql_helm_values.volumePermissions.securityContext.runAsUser` has no direct equivalent and therefore the volume permissions setup has to be disabled entirely in `postgresql_helm_values.volumePermissions.enabled`.
```

ccremer commented on docs/modules/ROOT/pages/how-tos/upgrade-9.x-to-10.x.adoc:
```diff
@@ -7,6 +7,10 @@ This guide describes the steps to perform an upgrade of the component from versi
+* `postgresql_helm_values.securityContext.enabled` changed to `postgresql_helm_values.primary.securityContext.enabled`.
+* `postgresql_helm_values.containerSecurityContext.enabled` changed to `postgresql_helm_values.primary.containerSecurityContext.enabled`.
+* `postgresql_helm_values.volumePermissions.securityContext.runAsUser` has no direct equivalent and therefore the volume permissions setup has to be disabled entirly `postgresql_helm_values.volumePermissions.enabled`.
+* `postgresql_helm_values.shmVolume.chmod.enabled` has no direct equivalent and therefore the shared volume setup has to be disabled entirly `postgresql_helm_values.shmVolume.enabled`.
```
```suggestion
* `postgresql_helm_values.shmVolume.chmod.enabled` has no direct equivalent and therefore the shared volume setup has to be disabled entirely in `postgresql_helm_values.shmVolume.enabled`.
```

megian commented:
> Might make sense to figure out if it's possible to configure sensible `securityPolicy` which doesn't cause PSA warnings, cf. https://kb.vshn.ch/oc4/explanations/pod_security.html

Can you elaborate what you exactly mean. Do you mean the postgres DB or Keycloak as well?

I think the DB security context is ok:
```
spec:
  containers:
  - name: postgresql
...
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      runAsUser: 1000990000
...
  initContainers:
  - command:
...
    name: copy-certs
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      runAsUser: 1000990000
...
  securityContext:
    fsGroup: 1000990000
    seLinuxOptions:
      level: s0:c31,c25
    seccompProfile:
      type: RuntimeDefault
```

simu commented:
> > Might make sense to figure out if it's possible to configure sensible `securityPolicy` which doesn't cause PSA warnings, cf. https://kb.vshn.ch/oc4/explanations/pod_security.html
> 
> Can you elaborate what you exactly mean. Do you mean the postgres DB or Keycloak as well?
> 
> I think the DB security context is ok:
> 
> ```
> spec:
>   containers:
>   - name: postgresql
> ...
>     securityContext:
>       allowPrivilegeEscalation: false
>       capabilities:
>         drop:
>         - ALL
>       runAsNonRoot: true
>       runAsUser: 1000990000
> ...
>   initContainers:
>   - command:
> ...
>     name: copy-certs
>     securityContext:
>       allowPrivilegeEscalation: false
>       capabilities:
>         drop:
>         - ALL
>       runAsNonRoot: true
>       runAsUser: 1000990000
> ...
>   securityContext:
>     fsGroup: 1000990000
>     seLinuxOptions:
>       level: s0:c31,c25
>     seccompProfile:
>       type: RuntimeDefault
> ```

As documented in the linked page, the Pod Security Admission check happens **before** OpenShift injects default security contexts, so the manifests resulting from compiling the component should already have reasonable `securityContext` settings, for example roughly the following for a non-privileged pod:

```
    securityContext:
      allowPrivilegeEscalation: false
      runAsNonRoot: true
      seccompProfile:
        type: RuntimeDefault
      capabilities:
        drop: ["ALL"]
        add: ["NET_BIND_SERVICE"]
```

megian commented:
> As documented in the linked page, the Pod Security Admission check happens **before** OpenShift injects default security contexts, so the manifests resulting from compiling the component should already have reasonable `securityContext` settings, for example roughly the following for a non-privileged pod:

Added the security context for the pod as well as for the container. As this seems not working on OpenShift 4.10, I've added a specific configuration for up to 4.10 and 4.11 and higher.

```
  Warning  FailedCreate      15s (x15 over 98s)  statefulset-controller  create Pod keycloak-postgresql-0 in StatefulSet keycloak-postgresql failed error: pods "keycloak-postgresql-0" is forbidden: unable to validate against any security context constraint: [pod.metadata.annotations.seccomp.security.alpha.kubernetes.io/pod: Forbidden: seccomp may not be set pod.metadata.annotations.container.seccomp.security.alpha.kubernetes.io/copy-certs: Forbidden: seccomp may not be set pod.metadata.annotations.container.seccomp.security.alpha.kubernetes.io/postgresql: Forbidden: seccomp may not be set provider "anyuid": Forbidden: not usable by user or serviceaccount provider "privileged-higher-prio": Forbidden: not usable by user or serviceaccount provider "cert-exporter-scc": Forbidden: not usable by user or serviceaccount provider "nonroot": Forbidden: not usable by user or serviceaccount pod.metadata.annotations.seccomp.security.alpha.kubernetes.io/pod: Forbidden: seccomp may not be set pod.metadata.annotations.container.seccomp.security.alpha.kubernetes.io/copy-certs: Forbidden: seccomp may not be set pod.metadata.annotations.container.seccomp.security.alpha.kubernetes.io/postgresql: Forbidden: seccomp may not be set provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount provider "elasticsearch-scc": Forbidden: not usable by user or serviceaccount provider "log-collector-scc": Forbidden: not usable by user or serviceaccount provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount provider "hostnetwork": Forbidden: not usable by user or serviceaccount provider "hostaccess": Forbidden: not usable by user or serviceaccount provider "syn-cluster-backup-etcd": Forbidden: not usable by user or serviceaccount provider "rook-ceph": Forbidden: not usable by user or serviceaccount provider "node-exporter": Forbidden: not usable by user or serviceaccount provider "rook-ceph-csi": Forbidden: not usable by user or serviceaccount provider "privileged": Forbidden: not usable by user or serviceaccount]
  ```

megian commented:
Added an issue for Keycloak itself #172.

simu reviewed (approved):
LGTM
